### DISCUSSION TITLE
Is it possible to use Django's test client with Reactivated?

### DISCUSSION THREAD

**Discussion by bartacc** [silviogutierrez/reactivated#366] [Q&A] [answered]
Hi,
I encountered an issue with [Django's test client](https://docs.djangoproject.com/en/5.0/topics/testing/tools/#the-test-client) when I tried to use it in my tests. Basically whenever I run a request to a view rendered by Reactivated, I get an error.

Steps to reproduce:

1. Create the example app with setup script from the [docs](https://www.reactivated.io/documentation/getting-started/)
2. Fill `tests.py` with the following:
```py
from django.test import TestCase

class MyTestCase(TestCase):
    def test_this_fails(self):
        self.client.get("")
```
3. Inside `nix-shell` run `./manage.py test`
4. Here is the output:

```
Found 1 test(s).
Creating test database for alias 'default'...
System check identified no issues (0 silenced).
node:internal/modules/cjs/loader:1080
  throw err;
  ^

Error: Cannot find module '/home/bart/Projects/Learning/Reactivated_Learning/my_app/node_modules/_reactivated/renderer.js'
    at Module._resolveFilename (node:internal/modules/cjs/loader:1077:15)
    at Module._load (node:internal/modules/cjs/loader:922:27)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:86:12)
    at node:internal/main/run_main_module:23:47 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}
```

Is it possible to configure Reactivated, so that this test client would work? The other solution which comes to my mind is to run tests in an environment where the webserver is running and then instead of Django's test client use something like [requests](https://requests.readthedocs.io/en/latest/) and communicate over actual HTTP on localhost. This is more of a hassle, so I'd like to avoid it, unless there is no other way.

Anyway, thanks for the good work on this library :)

**Comment by silviogutierrez** [comment]
Generally, I use `pytest`.

But for this to actually run and returned rendered responses, you'd need the React renderer to be running. There's no official way to do that yet, as you'd need to build the client.

If before running tests, you do `python manage.py build`, it'll build all the assets and `renderer.js` should exist. You have to do this before every run.

Another option is to set `REACTIVATED_SERVER = None` (this is undocumented) in your test settings. This will return just a JSON response from `self.client.get()` with all of your props. Good for testing everything *but* the React layer.

Give a try to those both. An "official" way to do e2e testing would be good.

**Reply by silviogutierrez** [reply]
@crgwbr may have his own approach.

**Reply by crgwbr** [reply, marked as answer]
We generally use the built-in Django testing tools rather than pytest. We have a few different kinds of tests:

1. Some e2e testing happens with Django's [LiveServerTestCase](https://docs.djangoproject.com/en/5.0/topics/testing/tools/#liveservertestcase) + a [Playwrite](https://playwright.dev/) headless browser.
2. Other testing uses the normal [TestCase](https://docs.djangoproject.com/en/5.0/topics/testing/tools/#testcase) class and the Django Test Client.

So, @bartacc, your approach should work. But, you do need to build the reactivated assets first. We accomplish that by means of a Makefile that looks something like this:

```Make
.PHONY: types
types:
	npm exec tsc --noEmit && \
	python manage.py generate_client_assets

.PHONY: build
build: types
	python manage.py build

.PHONY: test
test: build
	python manage.py test
```

**Reply by bartacc** [reply]
Thank you guys, both of these answers have really helped me. I got it working now :)
